### PR TITLE
Collection expressions: avoid intermediate List<T> if spread elements have known length

### PR DESCRIPTION
**Pull request by cston**
If all spread elements are countable *and the collection satisfies certain heuristics*, calculate the expected length of the resulting collection and:
- If the target is `T[]`, `Span<T>`, or `ReadOnlySpan<T>`, allocate the array or span at the expected length to avoid an intermediate buffer.
- If the target is `List<T>`, set the capacity to avoid resizing as items are added.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundCollectionExpression.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/BoundCollectionExpression.cs:20]:
```suggestion
                    {
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:636]: Consider naming this parameter for clarity.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:1919]: It's unclear to me from reading these comments what the difference between this and `ExpressionPlaceholder` is. Consider elaborating why we have both and what the difference between them is.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3519]: Why would this be default? Can it ever actually be `null`?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:392]: Consider making this static, and passing through `this` as a parameter, so the lambda can be cached.

**Review by 333fred — COMMENTED** [review]
Done review pass. Haven't looked at the tests in depth yet.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3519]: > Can it ever actually be `null`?

Are you asking if `node.Type` can be `null`? With this PR, the spread element type is always `null` since the `BoundExpression.Type` is not used.

It looks like we use the same `SetResultType(node, default)` for `BoundThrowExpression`, which seems like a similar case where the `BoundExpression.Type` is not meaningful.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: `ExpressionPlaceholder` represents the entire collection and `ElementPlaceholder` represents the current item in the iteration. Added a comment where `ElementPlaceholder` is used.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:411]: Not sure how I feel about this. Given we are creating an array I am not sure it is fine to create potentially very large numbers of temporaries just because a spread element (with runtime-known length) is present.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:411]: Basically, for this release, I would be absolutely fine with having an intermediate List for these cases. I could easily see a code generator wanting to write something like

```cs
public static readonly string[] arr1 = ["bunch", "of", "values", "like", "thousands"];
public static readonly string[] arr2 = [..arr1, "even", "more", "values"];
```

In this specific case, maybe blowing the stack on a static constructor is less likely, but still maybe possible.

Also, I think in some cases compiler can probably determine that a spread operator like `..arr1` is non-side-effecting to evaluate, e.g. we can read `Length` before evaluating the elements in-order without a problem, and optimize accordingly. I would def feel comfortable saying that if your Length property itself is side effecting, then bad things may happen, sorry.

**Review by RikkiGibson — APPROVED** [review]
The PR LGTM but it sounds like LDM may need to discuss the exact strategy in use here before we merge.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs]: Updated to only create temporaries up to the last spread element, and up to a maximum of 4 elements. That will cover cases such as `[..s, e1, e2, ...]` and `[..s1, ..s2]` in particular.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:992]: Added `SpreadElement_WritableReferenceLength`.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:596]: Ideally, we shouldn't require a temporary for an element that has a constant value.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4778]: We probably need a temp diagnostics bag that we add to `diagnostics` only if we're in the true path here. It looks like this will add use-site diagnostics from `Length` or `Count`, if any are found, even if it then returns false some some other reason.

File: src/Compilers/CSharp/Portable/BoundTree/BoundCollectionExpression.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/BoundCollectionExpression.cs:9]: I think a doc comment with an example of what `IncludingLastSpread` looks like would be helpful for future maintainability here.

**Review by 333fred — COMMENTED** [review]
Overall looking good. Would like some investigation of the `TryBindLengthOrCount` question before signing off though.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4778]: Good catch. I've added a test, `SpreadElement_LengthUseSiteError`, and I think we probably want to report the use-site errors (as it does now) even if `TryBindLengthOrCount` returns `false` because the reason it returned `false` might be due to the same issue that resulted in the use-site error. This is consistent with how `TryBindLengthOrCount` is used for patterns.

**Review by 333fred — APPROVED** [review]

**Review by RikkiGibson — APPROVED** [review]

**Comment by cston** [comment]
@333fred, @RikkiGibson, please review the latest commit which uses `List<T>..ctor(int capacity)`, thanks.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:147]: do you need to look to see if this constructor exists?  and only use it if so?  or is this ok to just do as is?

Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:147]: From the [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.-ctor?view=net-7.0#system-collections-generic-list-1-ctor(system-int32)), this constructor is available on .NET Framework 2.0 and later.

Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:147]: We're testing the missing constructor case in `KnownLength_List_MissingConstructor()`.

**Comment by cston** [comment]
> @333fred, @RikkiGibson, please review the latest commit which uses `List<T>..ctor(int capacity)`, thanks.

Moved the latest commit to the next PR: https://github.com/dotnet/roslyn/pull/70197

**Comment by RikkiGibson** [comment]
Is this PR ready to merge? @cston @jaredpar

**Comment by cston** [comment]
> Is this PR ready to merge? @cston @jaredpar

The changes in the PR have two approvals.

**Comment by jaredpar** [comment]
No, have to go through qb mode on these. Waiting for them to all be double reviewed beforetaking through QB

**Comment by jcouv** [comment]
Already merged elsewhere

**Comment by cston** [comment]
Merged as part of https://github.com/dotnet/roslyn/pull/70197.

**Pull request closed without merging**
